tools/clang/include/clang/SPIRV/Type.h

```diff
@@ -51,57 +57,66 @@ class Type {
   bool isCompositeType() const;
   bool isImageType() const;
 
-  static const Type *getType(SPIRVContext &ctx, spv::Op op,
-                             std::vector<uint32_t> arg = {},
-                             std::set<const Decoration *> decs = {});
-
-  static const Type *getVoid(SPIRVContext &ctx);
-  static const Type *getBool(SPIRVContext &ctx);
-  static const Type *getInt8(SPIRVContext &ctx);
-  static const Type *getUint8(SPIRVContext &ctx);
-  static const Type *getInt16(SPIRVContext &ctx);
-  static const Type *getUint16(SPIRVContext &ctx);
-  static const Type *getInt32(SPIRVContext &ctx);
-  static const Type *getUint32(SPIRVContext &ctx);
-  static const Type *getInt64(SPIRVContext &ctx);
-  static const Type *getUint64(SPIRVContext &ctx);
-  static const Type *getFloat16(SPIRVContext &ctx);
-  static const Type *getFloat32(SPIRVContext &ctx);
-  static const Type *getFloat64(SPIRVContext &ctx);
+  static const Type *getVoid(SPIRVContext &ctx, DecorationSet decs = {});
```

Comment:
Instead of requiring `DecorationSet` in the getter of each type, what about having a single `addDecoration{|Set}` method?

Not all the types need to be decorated; for those that do not need decorations, requiring `DecorationSet`s in the getter (which is essentially constructor) is not quite meaningful, even if we give them default values of empty sets.

My understanding of parameters in the constructor is the information we have to provide to make a valid object.  IHMO, `DecorationSet` is not such a parameter here.

Besides, the current implementation also does not have a way to construct the type first and then attach the decorations? We may have such use cases that we need to attach decorations gradually (I think)? Although I don't have a direct example here.